plugins/metrics/pkg/backend/rbac.go

```diff
@@ -0,0 +1,179 @@
+package backend
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	capabilityv1 "github.com/rancher/opni/pkg/apis/capability/v1"
+	v1 "github.com/rancher/opni/pkg/apis/core/v1"
+	"github.com/rancher/opni/pkg/capabilities/wellknown"
+	"github.com/rancher/opni/pkg/rbac"
+	"github.com/rancher/opni/pkg/storage"
+	"github.com/rancher/opni/pkg/util"
+	"github.com/rancher/opni/pkg/validation"
+	"github.com/rancher/opni/plugins/metrics/pkg/constants"
+	"github.com/rancher/opni/plugins/metrics/pkg/gateway/drivers"
+	metricsutil "github.com/rancher/opni/plugins/metrics/pkg/util"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type RBACBackend struct {
+	capabilityv1.UnsafeRBACManagerServer
+	RBACBackendConfig
+	util.Initializer
+}
+
+type RBACBackendConfig struct {
+	Logger         *zap.SugaredLogger `validate:"required"`
+	RoleStore      storage.RoleStore
+	StorageBackend storage.Backend
+}
+
+func (r *RBACBackend) Initialize(conf RBACBackendConfig) {
+	r.InitOnce(func() {
+		if err := metricsutil.Validate.Struct(conf); err != nil {
+			panic(err)
+		}
+		r.RBACBackendConfig = conf
+	})
+}
+
+func (r *RBACBackend) Info(_ context.Context, _ *emptypb.Empty) (*capabilityv1.Details, error) {
+	// Info must not block
+	return &capabilityv1.Details{
+		Name:    wellknown.CapabilityMetrics,
+		Source:  "plugin_metrics",
+		Drivers: drivers.ClusterDrivers.List(),
+	}, nil
+}
+
+func (r *RBACBackend) GetAvailablePermissions(_ context.Context, _ *emptypb.Empty) (*v1.AvailablePermissions, error) {
+	return &v1.AvailablePermissions{
+		Items: []*v1.PermissionDescription{
+			{
+				Type: string(v1.PermissionTypeCluster),
+				Verbs: []*v1.PermissionVerb{
+					v1.VerbGet(),
+				},
+			},
+		},
+	}, nil
+}
+
+func (r *RBACBackend) GetRole(ctx context.Context, in *v1.Reference) (*v1.Role, error) {
+	r.WaitForInit()
+	role, err := r.RoleStore.Get(ctx, in.GetId())
+	if err != nil {
+		return nil, err
+	}
+	return role, nil
+}
+
+func (r *RBACBackend) CreateRole(ctx context.Context, in *v1.Role) (*emptypb.Empty, error) {
+	if err := validation.Validate(in); err != nil {
+		return &emptypb.Empty{}, err
+	}
+
+	r.WaitForInit()
+
+	_, err := r.RoleStore.Get(ctx, in.Reference().GetId())
+	if err == nil {
+		return nil, storage.ErrAlreadyExists
+	}
+	if !storage.IsNotFound(err) {
+		return nil, err
+	}
+
+	err = r.RoleStore.Put(ctx, in.GetId(), in)
+	return &emptypb.Empty{}, err
+}
+
+func (r *RBACBackend) UpdateRole(ctx context.Context, in *v1.Role) (*emptypb.Empty, error) {
+	if err := validation.Validate(in); err != nil {
+		return &emptypb.Empty{}, err
+	}
+
+	r.WaitForInit()
+
+	oldRole, err := r.RoleStore.Get(ctx, in.Reference().GetId())
+	if err != nil {
+		return &emptypb.Empty{}, err
+	}
+
+	oldRole.Permissions = in.GetPermissions()
+	oldRole.Metadata = in.GetMetadata()
+	err = r.RoleStore.Put(ctx, oldRole.Reference().GetId(), oldRole)
```

Comment:
if the role has immutable fields, we might want to define a helper `apply(old, new)` function or `cmp(old, new) error` function, to prevent improper code changes down the line

Reply:
Or, perhaps a change to the  `rolestore` interface with an `Update` method that encapsulates the immutable field logic if it is shared between implementations